src/controlplane/scraper.go

```
@@ -0,0 +1,230 @@
+			continue
+		}
+
+		grouper := grouper.New(
+			s.deprecatedClient(component, pod.Name),
+			component.Queries,
+			s.logger,
+			pod.Name,
+		)
+
+		job := scrape.NewScrapeJob(string(component.Name), grouper, component.Specs)
+
+		s.logger.Debugf("Running job: %s", job.Name)
+
+		result := job.Populate(i, s.config.ClusterName, s.logger, s.k8sVersion)
+
+		if len(result.Errors) > 0 {
+			s.logger.Infof("Error populating data from %s: %v", job.Name, result.Error())
+		}
+	}
+
+	return nil
+}
+
+func (s *Scraper) deprecatedClient(c Component, podName string) client.HTTPClient {
+	timeout := 500 * time.Millisecond
+
+	authMethod := controlplaneClient.None
+	// Let mTLS take precedence over service account
+	switch {
+	case c.UseMTLSAuthentication:
+		authMethod = controlplaneClient.MTLS
+	case c.UseServiceAccountAuthentication:
+		authMethod = controlplaneClient.ServiceAccount
+	default:
+		authMethod = controlplaneClient.None
+	}
+
+	return controlplaneClient.New(
+		authMethod,
+		c.TLSSecretName,
+		c.TLSSecretNamespace,
+		s.logger,
+		s.K8s,
+		c.Endpoint,
+		c.SecureEndpoint,
+		s.config.NodeIP,
+		podName,
+		c.InsecureFallback,
+		timeout,
+	)
+}
+
+func (s *Scraper) discoverPod(c Component) (*corev1.Pod, error) {
+	var discoveredPod *corev1.Pod
+	// looks for the pod match a set of defined labels
+	for _, l := range c.Labels {
+		podLister, ok := s.PodListerByNamespace[c.Namespace]
+		if !ok {
+			return nil, fmt.Errorf("pod lister for namespace: %s not found for component: %s", c.Namespace, c.Name)
+		}
+
+		selector := labels.SelectorFromSet(labels.Set(l))
+		pods, err := podLister.List(selector)
+		if err != nil {
+			return nil, fmt.Errorf("fail to list pods for component:%v selector: %v", c.Name, l)
+		}
+
+		// Validation of returned pods.
+		for _, pod := range pods {
+			if pod.Spec.NodeName != s.config.NodeName {
+				s.logger.Debugf("discarding pod: %s running outside the node", pod.Name)
```

could we extract this to a function with a single return avoiding the continue and the break?

---

maybe also return an error or a bool to indicate no pod was found to avoid `discoveredPod == nil {`

---

i think this is more readeable now since there is no break any more , wdyt?